DataFormats/Detectors/TRD/include/DataFormatsTRD/TrapConfig3.h

```diff
@@ -0,0 +1,619 @@
+  uint32_t getRegisterValueByAddr(uint32_t regaddr, int detector, int rob, int mcm);
+
+  // get a registers name by addres and index;
+  std::string getRegNameByAddr(uint16_t addr);
+  std::string getRegNameByIdx(unsigned int regidx);
+
+  // get a registers index (enum value) by address or name
+  int32_t getRegIndexByAddr(unsigned int addr);
+  int32_t getRegIndexByName(const std::string& name);
+
+  // get a registers address by index or name
+  int32_t getRegAddrByIdx(unsigned int regidx);
+  int32_t getRegAddrByName(const std::string& name);
+
+  // retrieve the trap register info for a particular register, no need to index into det/rob/mcm as these are constant over a config
+  o2::trd::TrapRegInfo* getTrapRegInfoByAddr(uint32_t addr);
+  o2::trd::TrapRegInfo* getTrapRegInfoByIdx(uint32_t idx);
+
+  // get the indenfiying characteristics of a register given its address.
+  void getRegisterByAddr(uint32_t registeraddr, std::string& regname, int32_t& newregidx, int32_t& numberbits);
+
+  // return all the registers for a particular mcm
+  void getAllRegisters(int mcmidx, std::array<uint32_t, kLastReg>& registers);
+
+  // return all the mcm values for a particular register index
+  void getAllMCMByIndex(int regindex, std::array<uint32_t, o2::trd::constants::MAXMCMCOUNT>& registers);
+
+  // return all the mcm values for a particular register name
+  void getAllMCMByName(std::string registername, std::array<uint32_t, o2::trd::constants::MAXMCMCOUNT>& mcms);
+
+  // return all the mcm for a particular register address
+  void getAllMCMByAddress(int registeraddress, std::array<uint32_t, o2::trd::constants::MAXMCMCOUNT>& mcms);
+
+  // return all the config data in an unpacked array.
+  void getAll(std::array<uint32_t, kLastReg * o2::trd::constants::MAXMCMCOUNT>& configdata);
+
+  // return the full unpacked config event.
+  bool printRegister(int addr, int det, int rob, int mcm);
+  bool printRegister(TrapRegInfo* reg, int det, int rob, int mcm);
+
+  const std::array<o2::trd::TrapRegInfo, kLastReg>& getTrapRegisters() { return mTrapRegisters; }
+
+  // population pending
+  void setConfigVersion(uint32_t version) { mTrapConfigVersion = version; }           // these must some how be gotten from git or wingdb.
+  void setConfigNumber(uint32_t number) { mTrapConfigNumber = number; }               // these must be gotten from git or wingdb.
+  void setConfigSavedVersion(uint16_t version) { mTrapConfigSavedVersion = version; } // the version that is saved, for the ability to later save the config differently.
+  uint32_t getConfigVersion() { return mTrapConfigVersion; }                          // these must some how be gotten from git or wingdb.
+  uint32_t getConfigName() { return mTrapConfigNumber; }                              // these must be gotten from git or wingdb.
+  uint16_t getConfigSavedVersion() { return mTrapConfigSavedVersion; }                // the version that is saved, for the ability to later save the config differently.
```

Comment:
The config version and number can technically differ for different chambers. They should therefore not be a global property of the TRAPconfig. This feature led to the mixed configs in run 2, and should be avoided.
Also, the values should be retrieved from the registers for all the MCMs - at that point, you can check if they all agree, or if you have a mixed config.
What needs a lookup in git or wingDB is to translate the config number to the full name of the config, and the version to a git commit (at the moment it's just the svn release number).

Reply:
I thought mixed config was an error condition, that is handled on parsing and creating this class. Are you saying that we must allow for mixed configs? so configversion and name will be a vector of chambers long?

Reply:
I think here we have to clarify what we want to do with the TRAPconfigs. And right now I can think of three things:
- Make sure we get the correct settings into TRAPconfig. For this, mixed configs should be used as mixed configs, i.e. each MCM might have different settings.
- Issue a warning and mark the run when the config is mixed. This should be part of QC.
- Provide information what FEE config is/was active for a run. This should include the majority config, and which MCMs differed and what how their config was different.